### ISSUE TITLE
Broken and Outdated Doxygen Config

### ISSUE THREAD

**Issue by Levitanious** [cmangos/issues#1873]
### # The TL;DR Gist:
> For those in a hurry.

Turns out (surprise!) we have support for [automagical and blissful doxygen][doxy-conf] in the core.

It's _super_ neglected, somewhat outdated and whatnot -- but with just a small amount of fixing it can be brought back to life. It's amazing, important and useful. [Fix][patch-gist] and [improve][patch-gist-addon] it.

### # The Issue:
> I hope you're sitting tight, or you might be bored to death.

Back in [2008][old-tree-1], we still used [`autoconf`][ac-file] to prepare the build,
which was awesome and perfect until we moved to the other shit.

Moving to other shit happened right about after / during not-so-tasty-but-very-nasty times for the Project, so it comes as no surprise that we forgot to actually do something about the poor [`/doc/Doxyfile.in`][doxy-conf], `.gitignore` next to it... until 10 years later a Code Crawler meme disposed of my braincells and decided to check out the suspicious `/doc` folder.

So, i quickly grabbed a fresh `doxygen` package from my local dealer (hush-hush) and pointed it towards the what looked like a doxygen config file (and i wasn't wrong).

When the tool told me that some of the directives in the file were _deprecated_ i didn't expect what i was about to find out -- it's _**10 friggin' years old**_. It had to be a joke, right?

Wrong. Although it wasn't that terrible, just a quick `doxygen -u Doxyfile.in` and it should be running properly and nicely.

Right?

Wrong. Remember the [2008][old-tree-1] part of this story? This file has `.in` extension not just for lulz -- that's an include for `autoconf` with respective formatting and `@placeholders@` which don't work nicely with doxygen (because the actual config that should have been generated would be named Doxygen, it would be produced by `autoconf` tool and would be properly [ignored by git][gitignore]).

Now! Now, after cleaning that stuff up, checking it twice and rebooting my laptoppy three times just to be safe -- it ran and produced oh-so-much-welcomed output in html/ folder.

### # Additional stuff:
> You aren't tortured enough? I'LL HELP YOU OUT!

So, now that you actually know what happened (you've read the above? Nah, you didn't. But i dare you) -- i have a few proposals:

- Let's enable `EXTRACT_ALL`! It will allow us to get unstuck with manually maintained docs for much-much-much better CodeCrawling(TM) capability, also known as "API Autodocs".
- For the same reason, i suggest to turn on `EXTRACT_PRIVATE` and `EXTRACT_STATIC` -- much broader coverage for docs, much better indexing.
- Let's ditch old `DocStructure.dox`! It's not _very_ relevant, instead we should allow `RECURSIVE` search over entire `/src/` folder. Since we aren't capable of maintaining THAT much code at once, let's rely on autogenerated API summaries and add manual docs whenever we can on top of it.
- Let's enable `GENERATE_TREEVIEW`! It's literally 1000 times better than a _really_ old index page that still says "MaNGOS".
- To hell with custom-defined `FILE_PATTERNS`! Empty the variable, let doxygen do the smart scan!
- That `.gitignore` needs to be adjusted! We need only _one_ line there: `html/` -- for doxygen output folder. Rest doesn't happen anymore, and `html` without `/` makes git think you mean file instead, so the ignore wasn't working anyway. And instead of ignoring `Doxyfile` -- embrace it!
- As mentioned above: let's rename `Doxyfile.in` to `Doxyfile` to avoid any further confusion (we use `Doxyfile` for Recast, after all?)

### # Fix:
> That's what we're all about here -- fixing. Gotta reward someone who dared to read this all, right?
> ....right?

[There you go][patch-gist]! If you _really_ read eveything -- thank you, and i hope you've enjoyed it.

[Here][patch-gist-addon] is the patch with proposed settings, except renaming and `.gitignore` stuff.

### # Conclusion:
> You should've read TL;DR version, after all.

Am i allowed to post issues again, i wonder? :snake:  It was really fun~

[doxy-conf]: https://github.com/cmangos/mangos-tbc/blob/master/doc/Doxyfile.in
[patch-gist]: https://gist.github.com/Levitanious/628541656e45dac282ab8df8d4e05627
[old-tree-1]: https://github.com/cmangos/mangos-tbc/tree/b0299529f483171de4a23071c3dcfd708d92681f
[ac-file]: https://github.com/cmangos/mangos-tbc/blob/b0299529f483171de4a23071c3dcfd708d92681f/configure.ac
[gitignore]: https://github.com/cmangos/mangos-tbc/blob/master/doc/.gitignore
[patch-gist-addon]: https://gist.github.com/Levitanious/84bf26a6d9f929abb6c753f3204bff8a

**Comment by norcimoix** [comment]
I fully support getting DoxyGen up and running. I would actually really like to write a bunch of documentation. I need to get used to the core and figuring out and writing documentation for it is a great way to do that.

**Comment by esno** [comment]
hey @Levitanious how about integrating it into cmake? did a small POC this night and it needs some cleaning and error handling (https://github.com/blizzlike-org/cmangos-vanilla/commit/efb87b3af9d95e74535b1471bc5c4d10c7629c62) at least it currently installs into `CMAKE_INSTALL_PREFIX` which shouldn't happen on build stage, it runs only if boost and mysql-clients are installed and g3dlite will also built due to the way how the general CMakeFile.txt is written but that can be easily fixed
```cmake
...-vanilla/build > cmake ../ -DBUILD_DOCS=ON -DBUILD_EXTRACTORS=OFF -DBUILD_GAME_SERVER=OFF -DBUILD_LOGIN_SERVER=OFF -DCMAKE_INSTALL_PREFIX=./run
-- The C compiler identification is GNU 9.1.0
-- The CXX compiler identification is GNU 9.1.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done

-- This script builds the MaNGOS server.
  Options that can be used in order to configure the process:
    CMAKE_INSTALL_PREFIX    Path where the server should be installed to
    PCH                     Use precompiled headers
    DEBUG                   Include additional debug-code in core
    WARNINGS                Show all warnings during compile
    POSTGRESQL              Use PostgreSQL instead of mysql
    BUILD_GAME_SERVER       Build game server (core server)
    BUILD_LOGIN_SERVER      Build login server (auth server)
    BUILD_EXTRACTORS        Build map/dbc/vmap/mmap extractor
    BUILD_SCRIPTDEV         Build scriptdev. (Disable it to speedup build in dev mode by not including scripts)
    BUILD_PLAYERBOT         Build Playerbot mod
    BUILD_RECASTDEMOMOD     Build map/vmap/mmap viewer
    BUILD_GIT_ID            Build git_id

  To set an option simply type -D<OPTION>=<VALUE> after 'cmake <srcs>'.
  Also, you can specify the generator with -G. see 'cmake --help' for more details
  For example:
    Build server
    cmake -DCMAKE_INSTALL_PREFIX=../opt/cmangos ..

    Build full server with extractor
    cmake -DCMAKE_INSTALL_PREFIX=../opt/cmangos -DBUILD_EXTRACTORS=ON ..

    Build only server without scripts and extractors and in debug mode
    cmake -DCMAKE_INSTALL_PREFIX=../opt/cmangos -DDEBUG=ON -DBUILD_ALL_SCRIPTS=OFF ..

    Build only login server
    cmake -DCMAKE_INSTALL_PREFIX=../opt/cmangos -DBUILD_GAME_SERVER=OFF ..

-- Detected 64-bit platform
-- UNIX: Configuring uninstall target
-- UNIX: Created uninstall target
-- UNIX: Detected compiler: /usr/bin/cc
-- Performing Test COMPILER_SUPPORTS_CXX11
-- Performing Test COMPILER_SUPPORTS_CXX11 - Success
-- Performing Test COMPILER_SUPPORTS_CXX0X
-- Performing Test COMPILER_SUPPORTS_CXX0X - Success
-- GCC: SFMT enabled, SSE2 flags forced
Defaulting to release build.
-- Found Git: /usr/bin/git (found version "2.22.0") 
-- cotire 1.7.9 loaded.
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.69.0
-- Found the following Boost libraries:
--   system
--   program_options
--   thread
--   regex
--   chrono
--   date_time
--   atomic
-- Using mysql-config: /usr/bin/mysql_config
-- Found MySQL library: /usr/lib/libmariadb.so
-- Found MySQL headers: /usr/include/mysql
-- Found OpenSSL: /usr/lib/libcrypto.so (found version "1.1.1c")  
-- BUILD_SCRIPTDEV forced to OFF due to BUILD_GAME_SERVER is not set
-- CMaNGOS-Core revision : f0c4bb58e47edc0835be949ede3574bd6cb0cf13
-- Revision time stamp   : "2019-07-10T19:36:56+00:00"
-- Install server to     : /tmp/cmangos-vanilla/build/run
-- Use PCH               : Yes (default)
-- Build in debug-mode   : No  (default)
-- Build game server     : No
-- Build login server    : No
-- Build ScriptDev       : No
-- Build Playerbot       : No  (default)
-- Build extractors      : No  (default)
-- Build RecastDemoMod   : No  (default)
-- Build git_id          : No  (default)

-- Found Doxygen: /usr/bin/doxygen (found version "1.8.15") found components:  doxygen dot 
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/cmangos-vanilla/build
...-vanilla/build > make -j24                                                                   vanilla/bugfix/doxygen
Scanning dependencies of target doc
Scanning dependencies of target g3dlite
[  2%] Generating documentation with Doxygen
[  4%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/BinaryOutput.cpp.o
[  6%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/BinaryInput.cpp.o
[  9%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/BinaryFormat.cpp.o
[ 11%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/AABox.cpp.o
[ 13%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Box.cpp.o
[ 15%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Any.cpp.o
[ 18%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Capsule.cpp.o
[ 20%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/CollisionDetection.cpp.o
[ 22%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/CoordinateFrame.cpp.o
[ 25%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Crypto.cpp.o
[ 27%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Cylinder.cpp.o
[ 29%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/FileSystem.cpp.o
[ 31%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Matrix4.cpp.o
[ 34%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Log.cpp.o
[ 36%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Line.cpp.o
[ 38%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Matrix3.cpp.o
[ 40%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/MemoryManager.cpp.o
[ 43%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/LineSegment.cpp.o
[ 45%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/PhysicsFrame.cpp.o
[ 47%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Plane.cpp.o
[ 50%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Random.cpp.o
[ 52%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Quat.cpp.o
[ 54%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Ray.cpp.o
[ 56%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/ReferenceCount.cpp.o
[ 59%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/RegistryUtil.cpp.o
[ 61%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Sphere.cpp.o
[ 63%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/System.cpp.o
[ 65%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/TextInput.cpp.o
[ 68%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/TextOutput.cpp.o
[ 70%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Triangle.cpp.o
[ 72%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/UprightFrame.cpp.o
[ 75%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Vector2.cpp.o
[ 77%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Vector3.cpp.o
[ 79%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/Vector4.cpp.o
[ 81%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/debugAssert.cpp.o
[ 84%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/fileutils.cpp.o
[ 86%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/format.cpp.o
[ 88%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/g3dfnmatch.cpp.o
[ 88%] Built target doc
[ 90%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/g3dmath.cpp.o
[ 93%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/prompt.cpp.o
[ 95%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/stringutils.cpp.o
[ 97%] Building CXX object dep/src/g3dlite/CMakeFiles/g3dlite.dir/uint128.cpp.o
[100%] Linking CXX static library libg3dlite.a
[100%] Built target g3dlite
```

**Comment by Levitanious** [comment]
Closing it as resolved by @esno with https://github.com/cmangos/mangos-classic/pull/403
